### ISSUE TITLE
support for asymmetric learning rates (c/o distributional rl)

### ISSUE THREAD

**Issue by mobeets** [mobeets/vrnnstark#1]
To implement distributional RL, we need the following:

- have the ValueRNN's output be `K > 1` (trivial; set via `output_size`)
- now, the TD error will be `K` dimensional, but because we are using MSELoss as our loss function, this is no big deal

The difference, though, is that we want each of these outputs to have their own (asymmetric) learning rate. Specifically, value unit/output $i$ should have weight updates:

$\Delta V_i(t) = \alpha_i^+ f(\delta_i(t)) $
when $\delta_i(t) > 0$ and 
$\Delta V_i(t) = \alpha_i^- f(\delta_i(t)) $
when $\delta_i(t) ≤ 0$ 

where $\delta_i(t)$ is the RPE, and $f$ can be either the identity function or the sign function (yielding expectiles or quantiles)
(see Dabney et al., 2020)

So if $f$ is the identity function, the only change we need to implement is $\alpha_i^+$ and $\alpha_i^-$.

Currently, the learning rates are parameter-specific, and controlled via Adam.
What we need to implement (given a flag) is the following:

- choose the [SGD](https://pytorch.org/docs/stable/generated/torch.optim.SGD.html#torch.optim.SGD) optimizer, with no momentum
- also create "parameter groups" (info [here](https://stackoverflow.com/questions/73629330/what-exactly-is-meant-by-param-groups-in-pytorch)), which would be provided to SGD; we would want one parameter group per value output unit, so we can set each learning rate separately...

Or I dunno, this may not be the best way to do this.

**Comment by mobeets** [comment]
Some screenshots from the Methods of Dabney et al., 2020:

<img width="404" alt="Screenshot 2023-06-07 at 10 56 14 AM" src="https://github.com/mobeets/valuernn/assets/1677179/c9f1e7f1-9018-441b-943b-12bb6582037e">
<img width="401" alt="Screenshot 2023-06-07 at 10 56 22 AM" src="https://github.com/mobeets/valuernn/assets/1677179/5d38a5ec-6f10-4a2c-a1ab-60f4aa5db06a">
<img width="397" alt="Screenshot 2023-06-07 at 10 56 32 AM" src="https://github.com/mobeets/valuernn/assets/1677179/f6f6b7a9-185a-4d78-a7b8-b567ee9d5112">

**Comment by mobeets** [comment]
In fact, this might be as simple as keeping Adam around, and instead like [this](https://github.com/Kchu/DeepRL_PyTorch/blob/master/Distributional_RL/2_QR_DQN.py):

```python
u = q_target.detach() - q_eval # (m, N_QUANT, N_QUANT)
tau = torch.FloatTensor(QUANTS_TARGET).view(1, -1, 1) # (1, N_QUANT, 1)
# note that tau is for present quantile
if USE_GPU:
    tau = tau.cuda()
weight = torch.abs(tau - u.le(0.).float()) # (m, N_QUANT, N_QUANT)
loss = F.smooth_l1_loss(q_eval, q_target.detach(), reduction='none')
# (m, N_QUANT, N_QUANT)
loss = torch.mean(weight * loss, dim=1).mean(dim=1)
```

where

```python
N_QUANT = 200
# quantiles
QUANTS = np.linspace(0.0, 1.0, N_QUANT + 1)[1:]
QUANTS_TARGET = (np.linspace(0.0, 1.0, N_QUANT + 1)[:-1] + QUANTS)/2
```

Basically what this does is the following: `u` here is our RPE. And `tau` is our $\alpha^+$, and then `|tau-1|` is our $\alpha^-$. So if our RPE is positive, we have `weight = tau`, meaning our weight is $\alpha^+$; and if our RPE is negative we have `weight = |tau-1|`, meaning our weight is $\alpha^-$.
